src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo.java

```diff
@@ -0,0 +1,26 @@
+/*
+IllegalType
+validateAbstractClassNames = (default)false
+illegalClassNames = (default)HashMap, HashSet, LinkedHashMap, LinkedHashSet, TreeMap, TreeSet, \
+                    java.util.HashMap, java.util.HashSet, java.util.LinkedHashMap, \
+                    java.util.LinkedHashSet, java.util.TreeMap, java.util.TreeSet
+legalAbstractClassNames = (default)
+ignoredMethodNames = (default)getEnvironment, getInitialContext
+illegalAbstractClassNameFormat = (default)^(.*[.])?Abstract.*$
+memberModifiers = LITERAL_PUBLIC, LITERAL_PROTECTED, LITERAL_STATIC
+tokens = (default)ANNOTATION_FIELD_DEF, CLASS_DEF, INTERFACE_DEF, METHOD_CALL, METHOD_DEF, \
+         METHOD_REF, PARAMETER_DEF, VARIABLE_DEF, PATTERN_VARIABLE_DEF, RECORD_DEF, \
+         RECORD_COMPONENT_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.coding.illegaltype;
+
+import java.lang.String;
+import java.util.*;
+
+public record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
+        (HashSet productId, String identifier) {  // violation
```

Comment:
Record components are basically `private final` fields. Violations for the record components in this test case should be the same as the following for the given configs:

```
➜  javac Test1.java

➜  cat Test1.java
import java.util.HashSet;

public class Test1 {
    private final HashSet productId; // violation
    private final String identifier;

    public Test1(HashSet productId, String identifier) {
        this.productId = productId;
        this.identifier = identifier;
    }
}

➜  cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
        </module>
    </module>
</module>                                                                                                                 

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

➜  cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="FINAL"/>
        </module>
    </module>
</module>                                                                                                                

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

➜  cat config.xml                                           
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="FINAL, LITERAL_PRIVATE"/>
        </module>
    </module>
</module>%                                                                                                                 ➜  company java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

```

The config from this test case should not have any violations:

```

➜  cat config.xml                                           
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="LITERAL_PUBLIC, LITERAL_PROTECTED, LITERAL_STATIC"/>
        </module>
    </module>
</module>                                                                                                               

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
Audit done.

```

Reply:
@nmancus1 I have changed the access specifier from 
```
public record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
        (HashSet productId, String identifier) {  // violation
}
```

-->

```
private final record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
        (HashSet productId,  // violation
         String identifier) {  
}
```

Reply:
this is not exactly same as class fields.
Class fields can be HashSet but Ctor argument is Set:
```
public class Test1 {
    private final HashSet productId; // violation
    private final String identifier;

    public Test1(Set productId, String identifier) {
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
}
```
so such Class does not expose HashSet requirements outside, and do "bad" stuff completely internally. So it is fine. 

"records" do expose illegal type outside and almost force all user to use illegal type.

We need to work/behave as on records as we behave on constructors.
if 
```
    public Test1(HashSet productId, String identifier) { // violation
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
```
then
```
     record Test1(HashSet productId, String identifier); /// vilation
```

this should remove special treat consideration of FINAL and PRIVATE from code.

Reply:
@romani sir, I am getting no violation when I am using:
```
    public Test1(HashSet productId, String identifier) { // violation
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
```

Reply:
Okay Done.

Reply:
> We need to work/behave as on records as we behave on constructors.

This was my first thought as well, but this check does not check constructors parameters at all:

```java
    private void visitParameterDef(DetailAST parameterDef) {
        final DetailAST grandParentAST = parameterDef.getParent().getParent();

        if (grandParentAST.getType() == TokenTypes.METHOD_DEF && isCheckedMethod(grandParentAST)) {
            checkClassName(parameterDef);
        }
    }

```

Example:

```bash
        |--CTOR_DEF -> CTOR_DEF [4:4]
...
        |   |--PARAMETERS -> PARAMETERS [4:16]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [4:16]
...
        |--METHOD_DEF -> METHOD_DEF [8:4]
...
        |   |--PARAMETERS -> PARAMETERS [8:16]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [8:16]
...


```

This check relies on the given class's fields to put violations on illegal types, not constructor parameters. To me, this means that we should check record components as the fields they represent (`private final ...`), because this is how check currently operates on classes.

> so such Class does not expose HashSet requirements outside, and do "bad" stuff completely internally

This check does not have idea of "levels" of access modifiers; it only checks that if token has specified access modifiers (if `memberModifier` property is used) then we verify that it does not declare an illegal type. So, there is no concept of "outside" or "inside" to this check.

So, we have to make a decision on how to treat record components, and what implied modifiers that they carry. If we do not, they will not be checked at all if `memberModifiers` property is used at all. I can think of a few different solutions:

1. use example above (treat them as `private final`)
2. treat them as `public`
3. always check them, regardless of specified `memberModifiers` property as long as token is specified in config or default
4. let record components "inherit" access modifier of record declaration

Reply:
So Do I need to change anything? @nmancus1 @romani

Reply:
Do I need to revert changes again? @romani @nmancus1

Reply:
@romani sir, I updated the `SET in CTOR` and type cast `HashSet` in my recent changes.
https://github.com/checkstyle/checkstyle/pull/10958/files#diff-ca66a50cfe72a8a6786eac504d6efe1f7a3f6f79bb3428da3d07ae82330b4429R56-R58

Reply:
> but this check does not check constructor parameters at all

it is separate new issue.
Check is very general, it tries to catch illegal classes in all places. It does validation of Method arguments. 

Please see proposal to update javadoc at https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762465306 to make it clear that modifiers arenot considered for RECORD_COMPONENT_DEF

Reply:
@nmancus1 I think we can create a new issue regarding this?